File: examples/Algorithm/python/09_tree_bst.py
"""
Tree and Binary Search Tree (Tree & BST)
Tree and Binary Search Tree

Implements tree structures and BST operations.
"""

from typing import List, Optional, Generator
from collections import deque


# =============================================================================
# 1. Binary Tree Node
# =============================================================================

class TreeNode:
    """Binary tree node"""

    def __init__(self, val: int, left: 'TreeNode' = None, right: 'TreeNode' = None):
        self.val = val
        self.left = left
        self.right = right

    def __repr__(self):
        return f"TreeNode({self.val})"


# =============================================================================
# 2. Tree Traversal
# =============================================================================

def preorder_recursive(root: TreeNode) -> List[int]:
    """Preorder traversal (recursive) - O(n)"""
    result = []

    def traverse(node):
        if not node:
            return
        result.append(node.val)
        traverse(node.left)
        traverse(node.right)

    traverse(root)
    return result


def preorder_iterative(root: TreeNode) -> List[int]:
    """Preorder traversal (iterative) - O(n)"""
    if not root:
        return []

    result = []
    stack = [root]

    while stack:
        node = stack.pop()
        result.append(node.val)

        # Push right first (so left is processed first)
        if node.right:
            stack.append(node.right)
        if node.left:
            stack.append(node.left)

    return result


def inorder_recursive(root: TreeNode) -> List[int]:
    """Inorder traversal (recursive) - O(n)"""
    result = []

    def traverse(node):
        if not node:
            return
        traverse(node.left)
        result.append(node.val)
        traverse(node.right)

    traverse(root)
    return result


def inorder_iterative(root: TreeNode) -> List[int]:
    """Inorder traversal (iterative) - O(n)"""
    result = []
    stack = []
    current = root

    while stack or current:
        # Go to the leftmost node
        while current:
            stack.append(current)
            current = current.left

        current = stack.pop()
        result.append(current.val)
        current = current.right

    return result


def postorder_recursive(root: TreeNode) -> List[int]:
    """Postorder traversal (recursive) - O(n)"""
    result = []

    def traverse(node):
        if not node:
            return
        traverse(node.left)
        traverse(node.right)
        result.append(node.val)

    traverse(root)
    return result


def postorder_iterative(root: TreeNode) -> List[int]:
    """Postorder traversal (iterative) - O(n)"""
    if not root:
        return []

    result = []
    stack = [root]

    while stack:
        node = stack.pop()
        result.append(node.val)

        if node.left:
            stack.append(node.left)
        if node.right:
            stack.append(node.right)

    return result[::-1]  # Reverse


def level_order(root: TreeNode) -> List[List[int]]:
    """Level-order traversal (BFS) - O(n)"""
    if not root:
        return []

    result = []
    queue = deque([root])

    while queue:
        level_size = len(queue)
        level = []

        for _ in range(level_size):
            node = queue.popleft()
            level.append(node.val)

            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)

        result.append(level)

    return result


# =============================================================================
# 3. Binary Search Tree (BST)
# =============================================================================

class BST:
    """Binary Search Tree"""

    def __init__(self):
        self.root: Optional[TreeNode] = None

    def insert(self, val: int) -> None:
        """Insert node - Average O(log n), Worst O(n)"""
        if not self.root:
            self.root = TreeNode(val)
            return

        current = self.root
        while True:
            if val < current.val:
                if current.left is None:
                    current.left = TreeNode(val)
                    return
                current = current.left
            else:
                if current.right is None:
                    current.right = TreeNode(val)
                    return
                current = current.right

    def search(self, val: int) -> Optional[TreeNode]:
        """Search node - Average O(log n), Worst O(n)"""
        current = self.root

        while current:
            if val == current.val:
                return current
            elif val < current.val:
                current = current.left
            else:
                current = current.right

        return None

    def delete(self, val: int) -> bool:
        """Delete node - Average O(log n), Worst O(n)"""

        def find_min(node: TreeNode) -> TreeNode:
            while node.left:
                node = node.left
            return node

        def delete_recursive(node: TreeNode, val: int) -> Optional[TreeNode]:
            if not node:
                return None

            if val < node.val:
                node.left = delete_recursive(node.left, val)
            elif val > node.val:
                node.right = delete_recursive(node.right, val)
            else:
                # Found the node to delete

                # Case 1: Leaf node
                if not node.left and not node.right:
                    return None

                # Case 2: One child
                if not node.left:
                    return node.right
                if not node.right:
                    return node.left

                # Case 3: Two children - replace with successor (min of right subtree)
                successor = find_min(node.right)
                node.val = successor.val
                node.right = delete_recursive(node.right, successor.val)

            return node

        old_root = self.root
        self.root = delete_recursive(self.root, val)
        return old_root != self.root or (self.root and old_root.val != val if old_root else False)

    def inorder(self) -> List[int]:
        """Inorder traversal (sorted order)"""
        return inorder_recursive(self.root)

    def find_min(self) -> Optional[int]:
        """Find minimum value - O(h)"""
        if not self.root:
            return None

        current = self.root
        while current.left:
            current = current.left
        return current.val

    def find_max(self) -> Optional[int]:
        """Find maximum value - O(h)"""
        if not self.root:
            return None

        current = self.root
        while current.right:
            current = current.right
        return current.val


# =============================================================================
# 4. Tree Property Checks
# =============================================================================

def tree_height(root: TreeNode) -> int:
    """Calculate tree height - O(n)"""
    if not root:
        return -1  # Empty tree has height -1, single node has height 0

    return 1 + max(tree_height(root.left), tree_height(root.right))


def is_balanced(root: TreeNode) -> bool:
    """Check if tree is balanced - O(n)"""

    def check(node: TreeNode) -> int:
        if not node:
            return 0

        left_height = check(node.left)
        if left_height == -1:
            return -1

        right_height = check(node.right)
        if right_height == -1:
            return -1

        if abs(left_height - right_height) > 1:
            return -1

        return 1 + max(left_height, right_height)

    return check(root) != -1


def is_valid_bst(root: TreeNode) -> bool:
    """Check if tree is a valid BST - O(n)"""

    def validate(node: TreeNode, min_val: float, max_val: float) -> bool:
        if not node:
            return True

        if node.val <= min_val or node.val >= max_val:
            return False

        return (validate(node.left, min_val, node.val) and
                validate(node.right, node.val, max_val))

    return validate(root, float('-inf'), float('inf'))


def count_nodes(root: TreeNode) -> int:
    """Count nodes - O(n)"""
    if not root:
        return 0
    return 1 + count_nodes(root.left) + count_nodes(root.right)


# =============================================================================
# 5. Tree Transformation/Construction
# =============================================================================

def build_tree_from_list(values: List[Optional[int]]) -> Optional[TreeNode]:
    """Build tree from level-order list - O(n)"""
    if not values or values[0] is None:
        return None

    root = TreeNode(values[0])
    queue = deque([root])
    i = 1

    while queue and i < len(values):
        node = queue.popleft()

        if i < len(values) and values[i] is not None:
            node.left = TreeNode(values[i])
            queue.append(node.left)
        i += 1

        if i < len(values) and values[i] is not None:
            node.right = TreeNode(values[i])
            queue.append(node.right)
        i += 1

    return root


def sorted_array_to_bst(nums: List[int]) -> Optional[TreeNode]:
    """Build balanced BST from sorted array - O(n)"""
    if not nums:
        return None

    def build(left: int, right: int) -> Optional[TreeNode]:
        if left > right:
            return None

        mid = (left + right) // 2
        node = TreeNode(nums[mid])
        node.left = build(left, mid - 1)
        node.right = build(mid + 1, right)
        return node

    return build(0, len(nums) - 1)


def invert_tree(root: TreeNode) -> TreeNode:
    """Invert tree (mirror) - O(n)"""
    if not root:
        return None

    root.left, root.right = invert_tree(root.right), invert_tree(root.left)
    return root


# =============================================================================
# 6. Practical Problems
# =============================================================================

def lowest_common_ancestor(root: TreeNode, p: int, q: int) -> Optional[TreeNode]:
    """Lowest Common Ancestor (LCA) in BST - O(h)"""
    current = root

    while current:
        if p < current.val and q < current.val:
            current = current.left
        elif p > current.val and q > current.val:
            current = current.right
        else:
            return current

    return None


def kth_smallest(root: TreeNode, k: int) -> int:
    """Kth smallest value in BST - O(h + k)"""
    stack = []
    current = root
    count = 0

    while stack or current:
        while current:
            stack.append(current)
            current = current.left

        current = stack.pop()
        count += 1

        if count == k:
            return current.val

        current = current.right

    return -1


def path_sum(root: TreeNode, target: int) -> bool:
    """Check root-to-leaf path sum - O(n)"""
    if not root:
        return False

    if not root.left and not root.right:
        return root.val == target

    remaining = target - root.val
    return path_sum(root.left, remaining) or path_sum(root.right, remaining)


def serialize(root: TreeNode) -> str:
    """Serialize tree - O(n)"""
    if not root:
        return "[]"

    result = []
    queue = deque([root])

    while queue:
        node = queue.popleft()
        if node:
            result.append(str(node.val))
            queue.append(node.left)
            queue.append(node.right)
        else:
            result.append("null")

    # Remove trailing nulls
    while result and result[-1] == "null":
        result.pop()

    return "[" + ",".join(result) + "]"


# =============================================================================
# Utility: Tree Visualization
# =============================================================================

def print_tree(root: TreeNode, prefix: str = "", is_left: bool = True) -> None:
    """Print tree in ASCII"""
    if not root:
        return

    print(prefix + ("├── " if is_left else "└── ") + str(root.val))

    children = []
    if root.left:
        children.append((root.left, True))
    if root.right:
        children.append((root.right, False))

    for i, (child, is_left_child) in enumerate(children):
        extension = "│   " if is_left and i < len(children) - 1 else "    "
        print_tree(child, prefix + extension, is_left_child)


# =============================================================================
# Tests
# =============================================================================

def main():
    print("=" * 60)
    print("Tree and Binary Search Tree (Tree & BST) Examples")
    print("=" * 60)

    # 1. Tree Construction
    print("\n[1] Tree Construction")
    #       4
    #      / \
    #     2   6
    #    / \ / \
    #   1  3 5  7
    root = build_tree_from_list([4, 2, 6, 1, 3, 5, 7])
    print("    Level order: [4, 2, 6, 1, 3, 5, 7]")
    print("    Tree structure:")
    print_tree(root, "    ")

    # 2. Tree Traversal
    print("\n[2] Tree Traversal")
    print(f"    Preorder:  {preorder_recursive(root)}")
    print(f"    Inorder:   {inorder_recursive(root)}")
    print(f"    Postorder: {postorder_recursive(root)}")
    print(f"    Level:     {level_order(root)}")

    # 3. Tree Properties
    print("\n[3] Tree Properties")
    print(f"    Height: {tree_height(root)}")
    print(f"    Node count: {count_nodes(root)}")
    print(f"    Balanced: {is_balanced(root)}")
    print(f"    Valid BST: {is_valid_bst(root)}")

    # 4. BST Operations
    print("\n[4] BST Operations")
    bst = BST()
    for val in [5, 3, 7, 1, 4, 6, 8]:
        bst.insert(val)
    print(f"    Insert: [5, 3, 7, 1, 4, 6, 8]")
    print(f"    Inorder traversal: {bst.inorder()}")
    print(f"    Search 4: {bst.search(4)}")
    print(f"    Min: {bst.find_min()}, Max: {bst.find_max()}")

    bst.delete(3)
    print(f"    After delete 3: {bst.inorder()}")

    # 5. Sorted Array -> Balanced BST
    print("\n[5] Sorted Array -> Balanced BST")
    arr = [1, 2, 3, 4, 5, 6, 7]
    balanced_bst = sorted_array_to_bst(arr)
    print(f"    Input: {arr}")
    print(f"    Level order: {level_order(balanced_bst)}")

    # 6. LCA
    print("\n[6] Lowest Common Ancestor (LCA)")
    lca = lowest_common_ancestor(root, 1, 3)
    print(f"    LCA of nodes 1, 3: {lca.val if lca else None}")
    lca = lowest_common_ancestor(root, 1, 6)
    print(f"    LCA of nodes 1, 6: {lca.val if lca else None}")

    # 7. Kth Smallest Value
    print("\n[7] Kth Smallest Value")
    for k in [1, 3, 5]:
        print(f"    {k}th smallest: {kth_smallest(root, k)}")

    # 8. Path Sum
    print("\n[8] Root-to-Leaf Path Sum")
    print(f"    Sum 7 (4->2->1): {path_sum(root, 7)}")
    print(f"    Sum 10 (4->6): {path_sum(root, 10)}")

    # 9. Serialization
    print("\n[9] Tree Serialization")
    print(f"    {serialize(root)}")

    print("\n" + "=" * 60)


if __name__ == "__main__":
    main()

File: examples/Algorithm/c/09_tree_bst.c
/*
 * Tree and BST (Binary Search Tree)
 * Tree Traversal, BST Operations, Tree Properties
 *
 * Fundamental operations on tree data structures.
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

/* =============================================================================
 * 1. Binary Tree Node
 * ============================================================================= */

typedef struct TreeNode {
    int val;
    struct TreeNode* left;
    struct TreeNode* right;
} TreeNode;

TreeNode* create_node(int val) {
    TreeNode* node = malloc(sizeof(TreeNode));
    node->val = val;
    node->left = NULL;
    node->right = NULL;
    return node;
}

void free_tree(TreeNode* root) {
    if (root == NULL) return;
    free_tree(root->left);
    free_tree(root->right);
    free(root);
}

/* =============================================================================
 * 2. Tree Traversal
 * ============================================================================= */

void preorder(TreeNode* root) {
    if (root == NULL) return;
    printf("%d ", root->val);
    preorder(root->left);
    preorder(root->right);
}

void inorder(TreeNode* root) {
    if (root == NULL) return;
    inorder(root->left);
    printf("%d ", root->val);
    inorder(root->right);
}

void postorder(TreeNode* root) {
    if (root == NULL) return;
    postorder(root->left);
    postorder(root->right);
    printf("%d ", root->val);
}

/* Level-order traversal (BFS) */
void level_order(TreeNode* root) {
    if (root == NULL) return;

    TreeNode** queue = malloc(1000 * sizeof(TreeNode*));
    int front = 0, rear = 0;
    queue[rear++] = root;

    while (front < rear) {
        TreeNode* node = queue[front++];
        printf("%d ", node->val);

        if (node->left) queue[rear++] = node->left;
        if (node->right) queue[rear++] = node->right;
    }

    free(queue);
}

/* =============================================================================
 * 3. BST Operations
 * ============================================================================= */

TreeNode* bst_insert(TreeNode* root, int val) {
    if (root == NULL) return create_node(val);

    if (val < root->val)
        root->left = bst_insert(root->left, val);
    else if (val > root->val)
        root->right = bst_insert(root->right, val);

    return root;
}

TreeNode* bst_search(TreeNode* root, int val) {
    if (root == NULL || root->val == val)
        return root;

    if (val < root->val)
        return bst_search(root->left, val);
    return bst_search(root->right, val);
}

TreeNode* find_min(TreeNode* root) {
    while (root && root->left)
        root = root->left;
    return root;
}

TreeNode* find_max(TreeNode* root) {
    while (root && root->right)
        root = root->right;
    return root;
}

TreeNode* bst_delete(TreeNode* root, int val) {
    if (root == NULL) return NULL;

    if (val < root->val) {
        root->left = bst_delete(root->left, val);
    } else if (val > root->val) {
        root->right = bst_delete(root->right, val);
    } else {
        /* Node found */
        if (root->left == NULL) {
            TreeNode* temp = root->right;
            free(root);
            return temp;
        } else if (root->right == NULL) {
            TreeNode* temp = root->left;
            free(root);
            return temp;
        }

        /* Two children: replace with minimum of right subtree */
        TreeNode* successor = find_min(root->right);
        root->val = successor->val;
        root->right = bst_delete(root->right, successor->val);
    }

    return root;
}

/* =============================================================================
 * 4. Tree Properties
 * ============================================================================= */

int tree_height(TreeNode* root) {
    if (root == NULL) return 0;

    int left_h = tree_height(root->left);
    int right_h = tree_height(root->right);

    return 1 + (left_h > right_h ? left_h : right_h);
}

int count_nodes(TreeNode* root) {
    if (root == NULL) return 0;
    return 1 + count_nodes(root->left) + count_nodes(root->right);
}

int count_leaves(TreeNode* root) {
    if (root == NULL) return 0;
    if (root->left == NULL && root->right == NULL) return 1;
    return count_leaves(root->left) + count_leaves(root->right);
}

bool is_balanced(TreeNode* root) {
    if (root == NULL) return true;

    int left_h = tree_height(root->left);
    int right_h = tree_height(root->right);

    if (abs(left_h - right_h) > 1) return false;

    return is_balanced(root->left) && is_balanced(root->right);
}

/* BST validation */
bool is_bst_util(TreeNode* root, TreeNode* min_node, TreeNode* max_node) {
    if (root == NULL) return true;

    if (min_node && root->val <= min_node->val) return false;
    if (max_node && root->val >= max_node->val) return false;

    return is_bst_util(root->left, min_node, root) &&
           is_bst_util(root->right, root, max_node);
}

bool is_bst(TreeNode* root) {
    return is_bst_util(root, NULL, NULL);
}

/* =============================================================================
 * 5. Path Problems
 * ============================================================================= */

/* Root-to-leaf path sum */
bool has_path_sum(TreeNode* root, int target_sum) {
    if (root == NULL) return false;

    if (root->left == NULL && root->right == NULL)
        return target_sum == root->val;

    return has_path_sum(root->left, target_sum - root->val) ||
           has_path_sum(root->right, target_sum - root->val);
}

/* Maximum path sum */
int max_path_sum_util(TreeNode* root, int* max_sum) {
    if (root == NULL) return 0;

    int left = max_path_sum_util(root->left, max_sum);
    int right = max_path_sum_util(root->right, max_sum);

    left = left > 0 ? left : 0;
    right = right > 0 ? right : 0;

    int path_sum = root->val + left + right;
    if (path_sum > *max_sum) *max_sum = path_sum;

    return root->val + (left > right ? left : right);
}

int max_path_sum(TreeNode* root) {
    int max_sum = root ? root->val : 0;
    max_path_sum_util(root, &max_sum);
    return max_sum;
}

/* =============================================================================
 * 6. LCA (Lowest Common Ancestor)
 * ============================================================================= */

/* General tree LCA */
TreeNode* lca(TreeNode* root, int p, int q) {
    if (root == NULL) return NULL;
    if (root->val == p || root->val == q) return root;

    TreeNode* left = lca(root->left, p, q);
    TreeNode* right = lca(root->right, p, q);

    if (left && right) return root;
    return left ? left : right;
}

/* BST LCA */
TreeNode* lca_bst(TreeNode* root, int p, int q) {
    if (root == NULL) return NULL;

    if (p < root->val && q < root->val)
        return lca_bst(root->left, p, q);
    if (p > root->val && q > root->val)
        return lca_bst(root->right, p, q);

    return root;
}

/* =============================================================================
 * 7. Tree Transformations
 * ============================================================================= */

/* Mirror/Invert tree */
TreeNode* invert_tree(TreeNode* root) {
    if (root == NULL) return NULL;

    TreeNode* temp = root->left;
    root->left = invert_tree(root->right);
    root->right = invert_tree(temp);

    return root;
}

/* Tree serialization (preorder traversal) */
void serialize(TreeNode* root, int* arr, int* idx) {
    if (root == NULL) {
        arr[(*idx)++] = -1;  /* NULL marker */
        return;
    }
    arr[(*idx)++] = root->val;
    serialize(root->left, arr, idx);
    serialize(root->right, arr, idx);
}

TreeNode* deserialize(int* arr, int* idx, int n) {
    if (*idx >= n || arr[*idx] == -1) {
        (*idx)++;
        return NULL;
    }

    TreeNode* node = create_node(arr[(*idx)++]);
    node->left = deserialize(arr, idx, n);
    node->right = deserialize(arr, idx, n);
    return node;
}

/* =============================================================================
 * Test
 * ============================================================================= */

int main(void) {
    printf("============================================================\n");
    printf("Tree and BST Examples\n");
    printf("============================================================\n");

    /* Build BST */
    /*
     *        50
     *       /  \
     *      30   70
     *     / \   / \
     *    20 40 60 80
     */
    TreeNode* root = NULL;
    int values[] = {50, 30, 70, 20, 40, 60, 80};
    for (int i = 0; i < 7; i++) {
        root = bst_insert(root, values[i]);
    }

    /* 1. Traversal */
    printf("\n[1] Tree Traversal\n");
    printf("    Preorder: ");
    preorder(root);
    printf("\n");
    printf("    Inorder: ");
    inorder(root);
    printf("\n");
    printf("    Postorder: ");
    postorder(root);
    printf("\n");
    printf("    Level-order: ");
    level_order(root);
    printf("\n");

    /* 2. BST Operations */
    printf("\n[2] BST Operations\n");
    printf("    Search 40: %s\n", bst_search(root, 40) ? "found" : "not found");
    printf("    Search 45: %s\n", bst_search(root, 45) ? "found" : "not found");
    printf("    Minimum: %d\n", find_min(root)->val);
    printf("    Maximum: %d\n", find_max(root)->val);

    /* 3. Tree Properties */
    printf("\n[3] Tree Properties\n");
    printf("    Height: %d\n", tree_height(root));
    printf("    Node count: %d\n", count_nodes(root));
    printf("    Leaf count: %d\n", count_leaves(root));
    printf("    Balanced: %s\n", is_balanced(root) ? "yes" : "no");
    printf("    BST: %s\n", is_bst(root) ? "yes" : "no");

    /* 4. Path */
    printf("\n[4] Path Problems\n");
    printf("    Path sum 100 exists: %s\n", has_path_sum(root, 100) ? "yes" : "no");
    printf("    Maximum path sum: %d\n", max_path_sum(root));

    /* 5. LCA */
    printf("\n[5] LCA (Lowest Common Ancestor)\n");
    TreeNode* ancestor = lca_bst(root, 20, 40);
    printf("    LCA(20, 40): %d\n", ancestor ? ancestor->val : -1);
    ancestor = lca_bst(root, 20, 80);
    printf("    LCA(20, 80): %d\n", ancestor ? ancestor->val : -1);

    /* 6. Deletion */
    printf("\n[6] BST Deletion\n");
    printf("    Inorder before deletion: ");
    inorder(root);
    printf("\n");
    root = bst_delete(root, 30);
    printf("    Inorder after deleting 30: ");
    inorder(root);
    printf("\n");

    /* 7. Serialization */
    printf("\n[7] Tree Serialization/Deserialization\n");
    int serialized[100];
    int idx = 0;
    serialize(root, serialized, &idx);
    printf("    Serialized: ");
    for (int i = 0; i < idx; i++) {
        printf("%d ", serialized[i]);
    }
    printf("\n");

    int deserialize_idx = 0;
    TreeNode* restored = deserialize(serialized, &deserialize_idx, idx);
    printf("    Inorder after restore: ");
    inorder(restored);
    printf("\n");

    /* Cleanup */
    free_tree(root);
    free_tree(restored);

    printf("\n============================================================\n");

    return 0;
}
